### ISSUE TITLE
Any methods to accerlarate the running of fanc?

### ISSUE THREAD

**Issue by yanchunzhang** [vaquerizaslab/fanc#34]
Hi, 
I start use fanc recently and found it always take a much longer time than I expected. For example, when I try run fanc compare at whole genome level at 10kb resolution, the running time was estimated as 8 hours; and when I try use from-cooler to convert cooler to fan-c format, the software run for several hours and get almost nothing output...

Still I didn't find any multi-thread option for these process.

Is there any problem with my command line or do you have any suggetion or comments on this?
`/home/ch220811/.local/bin/fanc compare -l -I ${MATRIX1} ${MATRIX2} ${OUT}`
`/home/ch220811/.local/bin/fanc from-cooler ${PREFIX}.cool ${PREFIX}.fanc.hic`

Thank you!

**Comment by kaukrise** [comment]
Hi,

the commands you are using look fine to me.

The two applications you are describing are characterised by two central properties:

1. They are I/O-bound operations, because a matrix-like object is written to file in each case
2. They typically only need to be run once, and the output object is then used for further analyses

Writing of the output object (point 1) can only be done from a single thread, as per HDF5 specification - otherwise the file will be corrupted. While we could parallelise the remaining calculation (primarily the comparison part), the I/O operation is often the actual rate-limiting step. That aspect of these commands can be sped up by using an SSD.

Admittedly, though, we have not worked a lot on parallelising these commands. I have done numerous tests, however, that demonstrate that this is not as simple a task as it seems on the surface. For example, despite being allowed in principle, we have had corruptions and crashes attempting to read from the same opened HDF5 object in parallel. This can be prevented by using locks on the object, but that in turn defeats a lot of the reason for parallelising in the first place. Another issue is the way Python handles multiprocessing, which sends a copy of all necessary input data to each worker unless using very basic shared memory objects, which are not particularly suitable for some of the complexities of Hi-C data. I am also not an expert on this, so if you or anyone else has the know-how and wants to give parallelising these commands a go, feel free to [submit a pull request](https://github.com/vaquerizaslab/fanc/pulls)!

All of this being said, performance is a major focus for the `1.0.0` release. I have already re-written the low-level matrix retrieval algorithms to be much more efficient, and am seeing speedups upwards of 10x for the majority of FAN-C functions. The release is still in alpha, however, and requires more rigorous testing for different usage scenarios - bugs, mistakes, and other issues are to be expected when using it. If you are interested in contributing to the testing, please let me know, and I will set up a pre-release branch.

Cheers,

Kai
